config/module.go

```diff
@@ -233,72 +241,164 @@ const FirstRunSuccessSuffix = ".first_run_succeeded"
+	// that the first run script does not exist and we debug log and exit quietly.
+	firstRunPath, err := m.EvaluateFirstRunPath(unpackedModDir, logger)
+	if err != nil {
+		// TODO(RSDK-9067): some first run path evaluation errors should be promoted to WARN logs.
+		logger.Debugw("no first run script detected, skipping setup phase", "error", err)
+		return nil
+	}
+
+	logger = logger.WithFields("module", m.Name, "path", firstRunPath)
+	logger.Infow("executing first run script")
+
+	timeout := defaultFirstRunTimeout
+	if m.FirstRunTimeout > 0 {
+		timeout = m.FirstRunTimeout.Unwrap()
+	}
+	cmdCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	//nolint:gosec // Yes, we are deliberating executing arbitrary user code here.
+	cmd := exec.CommandContext(cmdCtx, firstRunPath)
+
+	cmd.Env = os.Environ()
+	for key, val := range env {
+		cmd.Env = append(cmd.Env, key+"="+val)
 	}
-	return "", noop, errors.New("no first run script")
+
+	stdOut, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	stdErr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	scanOut := bufio.NewScanner(stdOut)
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+
+		for scanOut.Scan() {
+			logger.Infow("got stdio", "output", scanOut.Text())
+		}
+		// This scanner keeps trying to read stdio until the command terminates,
+		// at which point the stdio pipe handle is no longer available. This sometimes
+		// results in an `os.ErrClosed`, which we discard.
+		if err := scanOut.Err(); err != nil && !errors.Is(err, os.ErrClosed) {
+			logger.Errorw("error scanning stdio", "error", err)
+		}
+	}()
+	scanErr := bufio.NewScanner(stdErr)
+	go func() {
+		defer wg.Done()
+
+		for scanErr.Scan() {
+			logger.Warnw("got stderr", "output", scanErr.Text())
+		}
+		// This scanner keeps trying to read stderr until the command terminates,
+		// at which point the stderr pipe handle is no longer available. This sometimes
+		// results in an `os.ErrClosed`, which we discard.
+		if err := scanErr.Err(); err != nil && !errors.Is(err, os.ErrClosed) {
+			logger.Errorw("error scanning stderr", "error", err)
+		}
+	}()
+	if err := cmd.Start(); err != nil {
+		logger.Errorw("failed to start first run script", "error", err)
+		return err
+	}
+	if err := cmd.Wait(); err != nil {
+		logger.Errorw("first run script failed", "error", err)
+		return err
+	}
+	wg.Wait()
+	logger.Info("first run script succeeded")
+
+	// Mark success by writing a marker file to disk. This is a best
+	// effort; if writing to disk fails the setup phase will run again
+	// for this module and version and we are okay with that.
+	//nolint:gosec // safe
+	markerFile, err := os.Create(firstRunSuccessPath)
+	if err != nil {
+		logger.Errorw("failed to mark success", "error", err)
+		return nil
+	}
+	if err = markerFile.Close(); err != nil {
+		logger.Errorw("failed to close marker file", "error", err)
+		return nil
+	}
```

Comment:
this fixes a bug in the original logic - I never closed the marker file after creating it :sob: